src/typechecker/map.lisp

```diff
@@ -0,0 +1,66 @@
+;;; Environment maps provide readable representations of hash
+;;; table-valued fields present in Coalton environment updates
+;;; in generated Lisp code.
+
+(defpackage #:coalton-impl/typechecker/map
+  (:use
+   #:cl)
+  (:local-nicknames
+   (#:util #:coalton-impl/util))
+  (:export
+   #:environment-map
+   #:make-map
+   #:get-table
+   #:get-value
+   #:set-value))
+
+(in-package #:coalton-impl/typechecker/map)
+
+(defstruct environment-map
+  "A readable, hash table-backed map."
+  (table (util:required 'table) :type hash-table :read-only t)
+  (test  (util:required 'test)  :type symbol     :read-only t))
+
+(defun make-map (&key initial-contents (test 'equal))
+  "Construct an environment map containing the keys and values of the association list ALIST.
+
+Environment maps have a readable representation as an alist.
+Hash table is initialized using TEST."
+  (let ((table (make-hash-table :test test)))
+    (dolist (cons initial-contents)
+      (alexandria:ensure-gethash (car cons) table (cdr cons)))
+    (make-environment-map :table table
+                          :test test)))
+
+(defmethod make-load-form ((self environment-map) &optional environment)
```

Comment:
@jbouwman This will make a load form for the object itself, but don't we retain the problem of not having a load form for the underlying hash table? This is a question out of ignorance; I'm not sure.

I'd have assumed we needed to handle the hash table specifically, since we were having trouble with the hash-tables to begin with, motivating this change in the first place.

Reply:
I believe the supported Lisps do already provide load forms for hash tables: otherwise the existing compilation scheme of macroexpansion inside compile-file wouldn't be working. This make-load-form method allows that to continue to work -- there are implementations of make-load-form for all other types that are reachable in persisted environments.

The print-object method that follows is the key to representing hash tables in portable, standalone lisp source code.

Reply:
Got it, thanks!